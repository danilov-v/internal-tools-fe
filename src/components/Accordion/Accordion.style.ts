import styled from 'styled-components';

type AccordionProps = {
  expanded: boolean;
};

export const Accordion = styled.header<AccordionProps>`
  overflow: hidden;
  max-height: ${({ expanded }) => (expanded ? '1500' : '0')}px;
  transition: max-height 0.5s;
  margin: 0;
`;

export const Content = styled.div<AccordionProps>`
  opacity: ${({ expanded }) => (expanded ? '1' : '0')};
  transition: opacity 0.3s;
`;
