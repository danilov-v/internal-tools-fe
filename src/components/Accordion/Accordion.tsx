import React, { useState, ReactElement } from 'react';

import { DropDownButton } from 'components/Button';
import * as S from './Accordion.style';

type AccordionProps = {
  isExpanded: boolean;
  render?: (toggle: () => void, isExpanded: boolean) => ReactElement;
  title?: string;
};

export const Accordion: React.FC<AccordionProps> = ({
  children,
  isExpanded,
  render,
  title,
}) => {
  const [expanded, setExpanded] = useState(isExpanded);
  const toggleExpanded = (): void => setExpanded(!expanded);

  return (
    <>
      {render ? (
        render(toggleExpanded, expanded)
      ) : (
        <DropDownButton onClick={toggleExpanded} expanded={expanded}>
          {title}
        </DropDownButton>
      )}
      <S.Accordion isExpanded={expanded}>{children}</S.Accordion>
    </>
  );
};
