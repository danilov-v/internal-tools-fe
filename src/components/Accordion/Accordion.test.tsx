import React from 'react';
import { fireEvent, render, RenderResult } from '@testing-library/react';

import { Accordion } from './Accordion';

const getComponent = (props = {}, content = ''): RenderResult => {
  const parsedProps = {
    isExpanded: true,
    ...props,
  };

  return render(<Accordion {...parsedProps}>{content}</Accordion>);
};

describe('Accordion component', () => {
  it('matches snapshots with required props only', () => {
    expect(getComponent().baseElement).toMatchSnapshot();
  });

  it('matches snapshot with additional props', () => {
    const component = getComponent({ title: 'title' });

    expect(component.baseElement).toMatchSnapshot();
  });

  it('expands by expand button click', () => {
    expect.assertions(2);

    const component = getComponent(
      { isExpanded: false, title: 'title' },
      'content',
    );

    const content = component.getByText('content');
    const button = component.getByText('title');

    expect(content).toHaveStyle('opacity: 0');

    fireEvent.click(button);

    expect(content).toHaveStyle('opacity: 1');
  });
});
