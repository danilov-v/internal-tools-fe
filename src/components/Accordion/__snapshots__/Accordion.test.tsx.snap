// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component matches snapshot with additional props 1`] = `
<body>
  <div>
    <button
      class="sc-AxjAm sc-AxirZ jwJGcr"
    >
      title
    </button>
    <div
      class="sc-AxhCb dwhDyV"
    />
  </div>
</body>
`;

exports[`Accordion component matches snapshots with required props only 1`] = `
<body>
  <div>
    <button
      class="sc-AxjAm sc-AxirZ jwJGcr"
    />
    <div
      class="sc-AxhCb dwhDyV"
    />
  </div>
</body>
`;
